azkaban-common/src/main/java/azkaban/project/Project.java

```diff
@@ -128,7 +131,7 @@ public Flow getFlow(final String flowId) {
   }
 
   public void setFlows(final Map<String, Flow> flows) {
-    this.flows = flows;
+    this.flows = new HashMap<>(flows);
```

Comment:
if you want them to be truly immutable, maybe consider this.flows =  Collections.unmodifiableMap(flows), same for jobPropsMap and propsList

Reply:
Actually, flows can be mutable. When uploading project in db, it will set the project id and project version for all the flows in the project. See azkaban.project.AzkabanProjectLoader#persistProject

Reply:
I'll change it to ImmutableMap since only the value inside flow object will be modified, but the map itself won't be modified. Thanks!